src/meshapi/views/panoramas.py

```diff
@@ -0,0 +1,170 @@
+from rest_framework.decorators import api_view, permission_classes
+from rest_framework.response import Response
+from rest_framework.views import status
+from meshapi.models import Building, Install
+from meshapi.permissions import HasPanoramaUpdatePermission
+
+from meshapi.util.django_pglocks import advisory_lock
+
+
+# Raised if we get total nonsense as a panorama title
+class BadPanoramaTitle(Exception):
+    pass
+
+
+# View called to make MeshDB refresh the panoramas.
+@api_view(["GET"])
+@permission_classes([HasPanoramaUpdatePermission])
+@advisory_lock("update_panoramas_lock")
+def update_panoramas_from_github(request):
+    # Check that we have all the environment variables we need
+    owner = os.environ.get("PANO_REPO_OWNER")
+    repo = os.environ.get("PANO_REPO")
+    branch = os.environ.get("PANO_BRANCH")
+    directory = os.environ.get("PANO_DIR")
+    host_url = os.environ.get("PANO_HOST")
+    token = os.environ.get("PANO_GITHUB_TOKEN")
+
+    if not owner or not repo or not branch or not directory or not host_url or not token:
+        return Response({"detail": "Did not find environment variables"}, status=status.HTTP_500_INTERNAL_SERVER_ERROR)
+
+    head_tree_sha = get_head_tree_sha(owner, repo, branch)
+    if not head_tree_sha:
+        return Response(
+            {"detail": "Could not get head tree SHA from GitHub"}, status=status.HTTP_500_INTERNAL_SERVER_ERROR
+        )
+
+    panorama_files = list_files_in_git_directory(owner, repo, directory, head_tree_sha)
+    if not panorama_files:
+        return Response({"detail": "Could not get file list from GitHub"}, status=status.HTTP_500_INTERNAL_SERVER_ERROR)
+
+    print(panorama_files)
+
+    panos = build_pano_dict(panorama_files)
+
+    panoramas_saved, warnings = set_panoramas(panos)
+
+    return Response(
+        {
+            "detail": f"Saved {panoramas_saved} panoramas. Got {len(warnings)} warnings.",
+            "saved": panoramas_saved,
+            "warnings": len(warnings),
+            "warn_install_nums": warnings,
+        },
+        status=status.HTTP_200_OK,
+    )
+
+
+def set_panoramas(panos: dict[str, list[str]]) -> tuple[int, list[str]]:
+    panoramas_saved = 0
+    warnings = []
+
+    host_url = os.environ.get("PANO_HOST")
+
+    for install_number, filenames in panos.items():
+        try:
+            install: Install = Install.objects.get(install_number=int(install_number))
+            panoramas = []
+            if not install:
+                print(
+                    f"Warning: Could not add panorama to building (Install #{install_number}). Install does not exist."
+                )
+                warnings.append(install_number)
+                continue
+            for filename in filenames:
+                file_url = f"{host_url}{filename}"
+                panoramas.append(file_url)
+            if install.building.panoramas == panoramas:
+                continue
+            install.building.panoramas = panoramas
```

Comment:
This overwrites earlier installs with the panos from later ones, which is not ideal, but is understandable. Could you just add a comment pointing this out and saying maybe we should fix this in the future?

Reply:
Oh, I guess we can drop data. Per my earlier reply, I'll see if I can create some logic to protect diffs and do something sensible.

Reply:
Well the cases are slightly different. Here, even if we start with a clean DB, this will cause earlier installs for the same building to get overwritten by later ones